src/platform/ExtendedOTARequestorDriver.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <platform/ExtendedOTARequestorDriver.h>
+#include <platform/OTARequestorInterface.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+namespace {
+
+constexpr System::Clock::Seconds32 kUserConsentPollInterval = System::Clock::Seconds32(30);
+
+} // namespace
+
+void ExtendedOTARequestorDriver::UpdateAvailable(const UpdateDescription & update, System::Clock::Seconds32 delay)
+{
+    VerifyOrDie(mRequestor != nullptr);
+
+    if (update.userConsentNeeded == true && mUserConsentDelegate)
+    {
+        bool localConfigDisabled = false;
+        VerifyOrdo(DeviceLayer::ConfigurationMgr().GetLocalConfigDisabled(localConfigDisabled) == CHIP_NO_ERROR,
+                   ChipLogProgress(SoftwareUpdate, "Failed to get local config disabled, using as false"));
+
+        if (localConfigDisabled == false)
+        {
+            chip::ota::UserConsentSubject subject;
+            CHIP_ERROR err = GetUserConsentSubject(subject, update);
+            if (err != CHIP_NO_ERROR)
+            {
+                ChipLogError(SoftwareUpdate, "Failed to get user consent subject");
+                return;
+            }
+
+            mDelayedActionTime = delay;
+            HandleUserConsentState(mUserConsentDelegate->GetUserConsentState(subject));
+            return;
+        }
+    }
+
+    ScheduleDelayedAction(
+        delay,
+        [](System::Layer *, void * context) { static_cast<ExtendedOTARequestorDriver *>(context)->mRequestor->DownloadUpdate(); },
+        this);
+}
+
+void ExtendedOTARequestorDriver::PollUserConsentState()
+{
+    HandleUserConsentState(mUserConsentDelegate->CheckDeferredUserConsentState());
+}
+
+CHIP_ERROR ExtendedOTARequestorDriver::GetUserConsentSubject(chip::ota::UserConsentSubject & subject,
+                                                             const UpdateDescription & update)
+{
+    // mLastUsedProvider has the provider fabric index and endpoint id
+    subject.fabricIndex        = mLastUsedProvider.fabricIndex;
+    subject.providerEndpointId = mLastUsedProvider.endpoint;
+
+    // TODO: As we cannot use the src/app/Server.h in here so, figure out a way to get the node id.
```

Comment:
Does the consent delegate on the requestor side really need the requestor node ID? Maybe it should be renamed to `peerNodeId`. Or maybe it doesn't makes sense to use a single structure for `UserConsentSubject` on both the provider and the requestor side. It doesn't need to be resolved in this PR, so I'm just throwing some ideas.